Add RSNLoader

mikelow opened the pull request:
This provides decompression for the rsn format used for spc collections. Rsn files are just rar archives, so this also adds support for rar files. I'm using the open source LPGL [unarr library](https://github.com/selmf/unarr) and opted to directly include the relevant source files, removing as much non-rar related code as possible. Alternatively, we could just import the project as a gitsubmodule 🤷‍♂️.

In terms of licensing, I did some research and I think we're in compliance. I've included the original license and an explaination in the lib/unarr/README.md with a note that this is a modified version of the code, a link to the original repo and reference to the unarr license. Yes, the original rar format itself has a restrictive license, but so long as unarr was clean-room reverse-engineered, and I expect it is given that it's based on the source code of popular MacOS app - [The Unarchiver](https://code.google.com/archive/p/theunarchiver/), everything should be hunky dory.

## Motivation and Context
rsn is a popular format for spc archives

## How Has This Been Tested?
Tested on MacOS x86/arm64 and Windows 10.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.

sykhro commented on src/main/loaders/RSNLoader.cpp:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  auto path = file->GetFullPath();
+  auto stream = ar_open_file(path.c_str());
```
I'm guessing there should be a `ar_close_file()` somewhere?

mikelow replied:
good catch 👍

sykhro commented:
> Alternatively, we could just import the project as a gitsubmodule 🤷‍♂️.

We won't update it often so I'd say it's okay to have it vendored.

I have to say, I'm not a fan of submodules for dependencies because they make life more complicated for CI (and sometimes for devs too, if you mix https and ssh in your environment). Ideally we'd want to use something like vcpkg, but we can't really afford a binary cache (you don't want to compile Qt locally) so it'd be more trouble than it's worth.

sykhro commented:
The `/usr/bin/ld: src/main/unarr/libunarr.a: error adding symbols: file format not recognized` error is likely due to LTO being used to compile `unarr`.

mikelow commented:
> The /usr/bin/ld: src/main/unarr/libunarr.a: error adding symbols: file format not recognized error is likely due to LTO being used to compile unarr.

Oh, you're way ahead of me then. As you can see, I've been going down a CI rabbit hole. I thought maybe we weren't setting up the llvm 17 linker correctly. You'll have to forgive me that I'm unfamiliar with LTO. I'll look into the libunarr CMake scripts.

> p.s.: I would like to keep clang 17 on CI because currently it has better codegen than AppleClang

Yeah, I was planning to try to get clang 17 working again on Mac. Reverting was part of my debugging process. Still, I'm not sure why this solved the Mac build. Would LTO still explain it?

mikelow commented:
> The /usr/bin/ld: src/main/unarr/libunarr.a: error adding symbols: file format not recognized error is likely due to LTO being used to compile unarr.

Brilliant! Fixed it. I've reverted all changes to build.yml.

sykhro commented:
> Would LTO still explain it?

Yes, `lld` from AppleClang handles LLVM object files slightly differently.  

> Brilliant! Fixed it

Ahh, yes, it's been almost 10 years and this is still a bug around CMake+lld :D

btw, it would be nice to turn on LTO for the whole program, but I think we have some undefined behaviour somewhere. I will set up ASAN+UBSAN sometime soon, so that we can run the program with those and try to squish memory corruption bugs

mikelow commented:
> btw, it would be nice to turn on LTO for the whole program, but I think we have some undefined behaviour somewhere. I will set up ASAN+UBSAN sometime soon, so that we can run the program with those and try to squish memory corruption bugs

This is all immensely educational for me. Glad to have your C++ knowledge.

mikelow commented:
@sykhro sorry that I seemed to have mistakenly merged this before approval. Let me know if you want me to make any changes and/or revert.
